pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserService.java

```diff
@@ -0,0 +1,129 @@
+        ActivityInstanceDao instanceDao = handle.attach(ActivityInstanceDao.class);
+        Set<Long> instanceIds = instanceDao.findAllInstanceIdsByUserIds(Collections.singleton(userId));
+        LOG.info("Found {} activity instances to delete", instanceIds.size());
+        DBUtils.checkDelete(instanceIds.size(), instanceDao.deleteAllByIds(instanceIds));
+        // Removing counters, events
+        handle.attach(JdbiEventConfigurationOccurrenceCounter.class).deleteAllByParticipantId(userId);
+        JdbiQueuedEvent queueEvent = handle.attach(JdbiQueuedEvent.class);
+        queueEvent.deleteQueuedNotificationSubstitutionsByUserIds(Collections.singleton(userId));
+        queueEvent.deleteQueuedNotificationsByUserIds(Collections.singleton(userId));
+        queueEvent.deleteQueuedEventsByUserIds(Collections.singleton(userId));
+        // Removing governances
+        UserGovernanceDao userGovernanceDao = handle.attach(UserGovernanceDao.class);
+        List<Governance> userGovernances = userGovernanceDao.findGovernancesByParticipantGuid(user.getGuid()).collect(Collectors.toList());
+        Set<String> studyGuids = userGovernances.stream()
+                .map(Governance::getGrantedStudies).flatMap(Collection::stream).map(GrantedStudy::getStudyGuid)
+                .collect(Collectors.toSet());
+        userGovernanceDao.deleteAllGovernancesForProxy(userId);
+        userGovernanceDao.deleteAllGovernancesForParticipant(userId);
+        handle.attach(StudyGovernanceSql.class).deleteAgeUpCandidateByParticipantId(userId);
+        // Removing user itself
+        handle.attach(DataExportDao.class).deleteDataSyncRequestsForUser(userId);
+        handle.attach(JdbiUser.class).deleteAllByIds(Collections.singleton(userId));
+
+        // Cleaning up the elasticsearch data
+        RestHighLevelClient esClient = ElasticsearchServiceUtil.getElasticsearchClient(ConfigManager.getInstance().getConfig());
+        BulkRequest bulkRequest = new BulkRequest().timeout("2m");
+        for (String studyGuid : studyGuids) {
+            StudyDto studyDto = handle.attach(JdbiUmbrellaStudy.class).findByStudyGuid(studyGuid);
+
+            String indexParticipant = ElasticsearchServiceUtil.getIndexForStudy(handle, studyDto,
+                    ElasticSearchIndexType.PARTICIPANTS);
+            String indexParticipantStructured = ElasticsearchServiceUtil.getIndexForStudy(handle, studyDto,
+                    ElasticSearchIndexType.PARTICIPANTS_STRUCTURED);
+
+            bulkRequest.add(new DeleteRequest()
+                    .index(indexParticipant)
+                    .type(REQUEST_TYPE)
+                    .id(userGuid));
+
+            bulkRequest.add(new DeleteRequest()
+                    .index(indexParticipantStructured)
+                    .type(REQUEST_TYPE)
+                    .id(userGuid));
+        }
+
+        BulkResponse bulkResponse = esClient.bulk(bulkRequest, RequestOptions.DEFAULT);
+
+        if (bulkResponse.hasFailures()) {
```

Comment:
What happens if we get a `404` as part of the bulk request? For example, a study may not have a `PARTICIPANTS` index and so the document is not found. Is that considered a failure? How do we handle that situation?

Reply:
In case of index absence, it will be a failure response. I assume we can't do anything in this case so I just logged this event.
In case of document absence it won't be a failure.
<img width="972" alt="image" src="https://user-images.githubusercontent.com/71830706/105346231-5d47c880-5bf6-11eb-82cf-6fcdc6a4cd99.png">

Reply:
Okay, in this case I think it's more common to see document absence, so I think this is fine for now.